feat(mdThemeProvider): register custom theme styles

Emeegeemee opened the pull request:
Closes #7708

Emeegeemee commented:
I signed it!

DevVersion commented on src/core/services/theming/theming.js:
```diff
@@ -153,6 +161,9 @@ function ThemingProvider($mdColorPalette) {
     extendPalette: extendPalette,
     theme: registerTheme,
 
+    registerStyles: function(styles) {
```
I initially thought, that there will be a function, which compiles the provided style on demand as well? Thoughts?

Emeegeemee replied:
Could you clarify what you mean by compile? I would expect the provided styles are valid CSS just like $MD_THEME_CSS and will just be appended to the constant before the themes are generated.

DevVersion replied:
@Emeegeemee Sorry, I got some different thoughts, which are not valid anymore, since `$mdColors` will come. 

Just thought again about your method, and it looks good.

ThomasBurleson commented:
*  Please provide a CodePen or Doc Theme Demo that justifies this feature addition. 
*  Also your test does not test anything. Please provide better tests.

> I am preparing to reject this PR as "too risky".

Emeegeemee commented:
I've updated the tests to be more specific and added a few missing checks, please let me know if you want something more specific tested.

As for a demo I'm not really sure where to start, from my perspective all of the currently implemented components serve as great demo's to the power of using the theming system. Would a document explaining how to create and register custom styles be better? I'm thinking of updating and moving [this guide](https://github.com/angular/material/blob/master/docs/guides/THEMES_IMPL_NOTES.md) into the theme section of the documentation if that make sense. 

As for your "too risky" comment could you give some examples that I could help reduce or mitigate. I definitely see the potential for developers making a mistake with registering styles. Doing so would cause the rest of the theming to break, but I believe that you would just get components without theming similar to if the themes were disabled.

ThomasBurleson commented:
@DevVersion - can you rebase this please and force push to this PR?

ThomasBurleson commented:
@DevVersion - ping!

DevVersion commented:
@Emeegeemee Thanks for the great PR again - We have a new PR, which includes your changes and gives deservedly credit to you :)
